src/lib/operators/join_sort_merge.cpp

```diff
@@ -247,35 +263,183 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractJoinOperatorImpl {
   }
 
   /**
-  * Emits all the combinations of row ids from the left table range and the right table range to the join output.
-  * I.e. the cross product of the ranges is emitted.
-  **/
-  void _emit_all_combinations(size_t output_cluster, TableRange left_range, TableRange right_range) {
+    * Emits all the combinations of row ids from the left table range and the right table range to the join output
+    * where also the secondary predicates are satisfied.
+    **/
+  void _emit_qualified_combinations(size_t output_cluster, TableRange left_range, TableRange right_range) {
+    if (_multi_predicate_join_evaluator) {
+      if (_mode == JoinMode::Inner) {
+        _emit_combinations_multi_predicated_inner(output_cluster, left_range, right_range);
+      } else if (_mode == JoinMode::Left) {
+        _emit_combinations_multi_predicated_left_outer(output_cluster, left_range, right_range);
+      } else if (_mode == JoinMode::Right) {
+        _emit_combinations_multi_predicated_right_outer(output_cluster, left_range, right_range);
+      } else if (_mode == JoinMode::FullOuter) {
+        _emit_combinations_multi_predicated_full_outer(output_cluster, left_range, right_range);
+      }
+    } else {
+      // no secondary join predicates
+      left_range.for_every_row_id(_sorted_left_table, [&](RowID left_row_id) {
+        right_range.for_every_row_id(_sorted_right_table, [&](RowID right_row_id) {
+          _emit_combination(output_cluster, left_row_id, right_row_id);
+        });
+      });
+    }
+  }
+
+  /**
+   * Only for multi predicated inner joins.
+   * Emits all the combinations of row ids from the left table range and the right table range to the join output
+   * where the secondary predicates are satisfied.
+   **/
+  void _emit_combinations_multi_predicated_inner(size_t output_cluster, TableRange left_range, TableRange right_range) {
     left_range.for_every_row_id(_sorted_left_table, [&](RowID left_row_id) {
       right_range.for_every_row_id(_sorted_right_table, [&](RowID right_row_id) {
-        _emit_combination(output_cluster, left_row_id, right_row_id);
+        if (_multi_predicate_join_evaluator->satisfies_all_predicates(left_row_id, right_row_id)) {
+          _emit_combination(output_cluster, left_row_id, right_row_id);
+        }
       });
     });
   }
 
   /**
-  * Emits all combinations of row ids from the left table range and a NULL value on the right side to the join output.
+  * Only for multi predicated left outer joins.
+  * Emits all the combinations of row ids from the left table range and the right table range to the join output
+  * where the secondary predicates are satisfied.
+  * For a left row id without a match, the combination [left row id|NULL row id] is emitted.
   **/
-  void _emit_right_null_combinations(size_t output_cluster, TableRange left_range) {
+  void _emit_combinations_multi_predicated_left_outer(size_t output_cluster, TableRange left_range,
```

Comment:
I trust that what seems to be very similar code in _emit_combinations_multi_predicated_left_outer, _emit_combinations_multi_predicated_right_outer and _emit_combinations_multi_predicated_full_outer cannot be avoided?

Reply:
Looks like only the ranges are swapped, correct? Should be straightforward to merge?

Reply:
That was my first impression too. I wasn't able to find an elegant solution.